# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
lambda x: x+2
(lambda x: x+2)(3)

5

#### Saving to a Variable

In [2]:
f = lambda x,y: x+y+2
f(3,5)

10

#### Multiple Inputs

In [3]:
(lambda x,y,z: x*z*y)(3,5,8)

#asorted as a variable
x = lambda x,y,z: x*y*z
x(3,5,8)

120

#### Passing a Lambda into a Function

In [4]:
def multiply(f,num):
    return f(num)

multiply(lambda x: x*x,4)

16

#### Returning a Lambda from a Function

In [8]:
def return_func():
    def multiply(num):
        return num*2
    return multiply
#return_func ---> mutiply
return_func()(2)

def return_lamb(b,c):
    return lambda x,a: x+a+b+c
print(return_lamb(1,2)(4,5))
    # b,c x,a
f = return_lamb(1,2)
f(4,5)

12


12

#### If Statements within Lambdas

In [10]:
f = lambda num: num*2 if num < 10 else num + 2
print(f(10))
print(f(8))

12
16


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [11]:
f = lambda num: num**3
print(f(3))

27


In [12]:
sorted([2,7,324,5,2,1])

[1, 2, 2, 5, 7, 324]

In [13]:
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [20]:
#using lmbda with sort to sort a list of dictionaries by a key from highest to lowest
k = {
    "name":"Kevin",
    "age":36
}
j = {
    "name":"John",
    "age":41
}
b = {
    "name":"Brian",
    "age":29
}
my_list = [k,j,b]
print(my_list)

print(sorted(my_list,key=lambda person_dict: person_dict["age"]))
print(sorted(my_list,key=lambda person_dict: person_dict["age"], reverse = True))

[{'name': 'Kevin', 'age': 36}, {'name': 'John', 'age': 41}, {'name': 'Brian', 'age': 29}]
[{'name': 'Brian', 'age': 29}, {'name': 'Kevin', 'age': 36}, {'name': 'John', 'age': 41}]
[{'name': 'John', 'age': 41}, {'name': 'Kevin', 'age': 36}, {'name': 'Brian', 'age': 29}]


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [21]:
def squared(um, num2):
    if num<10 and num2 < 10:
        return num**2, num2**2
    else: return num, num2
    
numbers = [4,10,3,2,6]
more_nums = [4,11,20,3,15]

#### Using Lambda's with Map

In [22]:
squared_nums = list(map(lambda x: x**2 if x<10 else x,numbers))
print(squared_nums)

[16, 10, 9, 4, 36]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [25]:
nums1=[4,9,6,5]
nums2=[8,9,3,4]
results = list(map(lambda x1, x2: (x1*2-1, x2*2-1), nums1, nums2))
results

[(7, 15), (17, 17), (11, 5), (9, 7)]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [26]:
names = ['Bob', 'Andy', 'Max', 'Evan', 'Angelia']
# names starting with 'A'
def a_names(name):
    if name[0].lower()=='a':
        return True
    else:
        return False
    
new_names_list = list(filter(a_names, names))
new_names_list

['Andy', 'Angelia']

#### Using Lambda's with Filter()

In [30]:
new_names_lamb = list(filter(lambda name: True if name[0].lower()=='a' else False, names))
new_names_lamb

['Andy', 'Angelia']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [35]:
from statistics import mean
numbers = [5,9,3,6,7,8.9,6.7,3.4,2.9]
results = list(filter(lambda num: True if num < mean(numbers) else False, numbers))
results

[5, 3, 3.4, 2.9]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [36]:
from functools import reduce
# reduce(function, iterable)

list1 = [9,5,7,5,8]

def substract_nums(num1, num2):
    return num1 - num2
result = reduce(substract_nums, list1)
result

-16

#### Using Lambda's with Reduce()

In [37]:
list3 = [1,5,2,5,3]

result_lamb = reduce(lambda x,y: x-y, list3)
print(result_lamb)

-14


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [39]:
list3 = [1,5,2,5,3]

multiply = reduce(lambda x,y: x*y, list3)
multiply

150

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [112]:
def recursive(i):
    print("hi")
    if i == 4:
        return
    i += 1
    recursive(i)
    
recursive(0)

hi
hi
hi
hi
hi


#### Implementing a Base Case

In [116]:
def add_nums(num):
    #set Base case
    if num <=1:
        print("iteration(1) = 1")
        return num
    else:
        print(f"iteration({num}) = {num + num - 1}")
        return num + add_nums(num-1)
add_nums(5)

iteration(5) = 9
iteration(4) = 7
iteration(3) = 5
iteration(2) = 3
iteration(1) = 1


15

#### Writing a Factorial Function

In [ ]:
def factorial(num):
    #set base case:
    if num <=1:
        return 1
    else:
        return num * factorial(num-1)
factorial()

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [41]:
def func(num):
    if num <=1:
        return 1
    else:
        return num - func(num-1)
func(5)

3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [42]:
def my_range(stop, start, step =1):
    while start < stop:
        yield start #function will store what is calculated inside the function
        start += step
for i in my_range(20,2):
    print(i)

2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [45]:
def squared(stop, start, step=1):
    while start > stop:
        yield start**2
        start -= step
for num in squared(0,10):
    print(num)

100
81
64
49
36
25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [47]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

def names(string):
    string=string.strip()
    if len(string) > 0:
        return True
    else:
        return False
    
result = list(filter(names, places))
result

['Argentina', 'San Diego', 'Boston', 'New York']

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [58]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

result = sorted(author, key=(lambda name: name.lower().split()[len(name.split())-1]))
print(result)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [86]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

# split the list of tuples into 02 lists
l1,l2 = zip(*places)

result = list(map(lambda x,y: (x,(9/5)*y + 32), l1,l2))
result

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [133]:
# Calculate fib function:
def fib_cal(num):
    if num <=1:
        return 1
    else:
        return (fib_cal(num-1) + fib_cal(num-2))

#Display fibonacci sequence
def fib(num):
    for i in range(num+1):
        print(f'Iteration {i}: {fib_cal(i)}')
fib(5)

Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
